### PR TITLE
adding bce with logits loss function

### PR DESCRIPTION
**Pull request by mrityunjay-tripathi**
Hello, Everyone
I am adding Binary Cross Entropy with Logits Loss function. Generally, it is used to combine the Sigmoid layer and Binary Cross-Entropy Loss in one class. This is numerically more stable than using a Sigmoid layer and a Binary Cross-Entropy layer explicitly as it prevents from imploding small values.

### CODE REVIEW

File: src/mlpack/tests/loss_functions_test.cpp [hunk 1]
```diff
@@ -33,6 +33,30 @@ using namespace mlpack::ann;
 
 BOOST_AUTO_TEST_SUITE(LossFunctionsTest);
 
+/**
+ * Binary Cross Entropy with Logits Loss function test.
+ */
+BOOST_AUTO_TEST_CASE(BCEWithLogitsLossTest)
+{
+  arma::mat input, output, target;
+  BCEWithLogitsLoss<> module;
```
Comment (himanshupathak21061998) [inline src/mlpack/tests/loss_functions_test.cpp]: Hey @mrityunjay-tripathi , You should `#include` your loss function because you compiler is not able to identify `BCEWitLogitsLoss` .

Comment (mrityunjay-tripathi) [reply src/mlpack/tests/loss_functions_test.cpp]: ohh XD

**Comment by mrityunjay-tripathi** [comment]
Hello @zoq @rcurtin 
The build failure in Windows VS14 Plain seems to be unrelated to this pull request, however, that fail is coming every time. Any reasons?

File: src/mlpack/methods/ann/loss_functions/bce_with_logits_loss.hpp [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+>
+class BCEWithLogitsLoss
+{
+ public:
+  /**
+   * Create the BCEWithLogitsLoss object.
+   *
+   * @param weight The manual rescaling factor given to loss.
+   * @param posWeight The weight of positive of examples.
+   *    posWeight = total negative examples / total positive examples.
```
Comment (birm) [inline src/mlpack/methods/ann/loss_functions/bce_with_logits_loss.hpp]: This almost reads like it's required that posWeight be set as such. I'm not sure how to convey this succinctly, but we could describe this as describing a good initial value for posWeight. Though, maybe only I'm reading it like this.

**Comment by birm** [comment]
@mrityunjay-tripathi No, the failure of the test on Visual Studio has no relevance to your proposed changes. I've added #2221 describing this issue in case you're curious.

File: src/mlpack/methods/ann/loss_functions/bce_with_logits_loss.hpp [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+   * @param reduction The boolean value, when 1, means reduction type is 'mean'
+   *    else if it is 0 then reduction type is 'sum'.
+   */
+  BCEWithLogitsLoss(
+    const double weight = 1.0,
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/bce_with_logits_loss.hpp]: Picky style issue, the first parameter could come right after the opening `(`.

File: src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+{
+  double lossThis;
+  double totalLossReduced = 0;
+  double m = input.max();
+  double logSigmoidX;
+  double logOneMinusSigmoidX;
+
+  for (size_t i = 0; i < input.n_elem; ++i)
+  {
+    logSigmoidX = m - std::log(std::exp(m) + std::exp(m - input[i]));
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp]: Maybe we should use `const double` here and for the `logOneMinusSigmoidX` parameter to make it clear they aren't changed.

File: src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+    logOneMinusSigmoidX = m - std::log(std::exp(m) + std::exp(m + input[i]));
+
+    lossThis =  - weight * (posWeight * target[i] * logSigmoidX
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp]: You could directly use, `totalLossReduced +=` here.

File: src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+    OutputType&& output)
+{
+  output.set_size(size(input));
+  double sigmoidX;
+
+  for (size_t i = 0; i < output.n_elem; ++i)
+  {
+    sigmoidX = 1 / (1 + std::exp(-input[i]));
+
+    output[i] = - weight * (posWeight * target[i] * (1 - sigmoidX)
```
Comment (zoq) [inline src/mlpack/methods/ann/loss_functions/bce_with_logits_loss_impl.hpp]: Should be easy to vectorize uisng armadillo point-wise multiplication, etc. the output calculation.

**Comment by rcurtin** [comment]
This code touches the ANN code, and there was recently a big refactoring of the ANN code in #2259, so be sure to merge the master branch into your branch here to make sure that nothing will fail if this PR is merged. :+1: (I'm pasting this message into all possibly relevant PRs.)

**Comment by kartikdutt18** [comment]
Hey there, This code touches the loss function portion of mlpack's codebase. Recently #2339 was merged where the switch was made to templated return type instead of just using double. So before we merge this, kindly make the changes for the same. I'm pasting this in all relevant PRs. Thanks a lot 👍
